### PR TITLE
Implement synchronized BiMap. Closes #25.

### LINKED ISSUE #25: Implement synchronized BiMaps.
The API `MutableBiMap<K, V>.asSynchronized()` already exists, but [the implementation](https://github.com/eclipse/eclipse-collections/blob/7.0.0/eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/AbstractMutableBiMap.java#L385) in `AbstractMutableBiMap` is unsupported.

### PR DESCRIPTION
**Pull request by bhav0904**

### CODE REVIEW

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/AbstractMutableBiMap.java [hunk 1]
```diff
@@ -543,6 +545,10 @@ public MutableShortBag collectShort(ShortFunction<? super V> shortFunction)
         return this.delegate.collectIf(predicate, function);
     }
 
+    /**
+     * @deprecated in 7.1. Use {@link OrderedIterable#zip(Iterable)} instead.
+     */
+    @Deprecated
```
Comment (itohro) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/AbstractMutableBiMap.java]: It looks odd to be included in the same commit. Could this be in a separate commit? Version number should be 8.0 now that 7.1 is already released. Also the link should be `zipWithIndex(Iterable)` not `zip(Iterable)` I think.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/AbstractMutableBiMap.java [hunk 2]
```diff
@@ -558,6 +564,10 @@ public MutableShortBag collectShort(ShortFunction<? super V> shortFunction)
         return this.delegate.groupByEach(function, new UnifiedSetMultimap<VV, V>());
     }
 
+    /**
+     * @deprecated in 7.1. Use {@link OrderedIterable#zip(Iterable)} instead.
+     */
+    @Deprecated
```
Comment (itohro) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/AbstractMutableBiMap.java]: Same as above, let's separate this to another commit. Update version# to 8.0, too.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/SynchronizedMutableBiMap.java [hunk 3]
```diff
@@ -0,0 +1,482 @@
...
+    }
+
+    protected SynchronizedMutableBiMap(MutableBiMap<K, V> delegate, Object lock)
+    {
+        super(delegate, lock);
+    }
+
+    /**
+     * This method will take a MutableBiMap and wrap it directly in a SynchronizedMutableMap.  It will
+     * take any other non-GS-map and first adapt it will a MapAdapter, and then return a
```
Comment (itohro) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/SynchronizedMutableBiMap.java]: Let's say "non-EC-map" instead of "non-GS-map".

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/SynchronizedMutableBiMap.java [hunk 4]
```diff
@@ -0,0 +1,482 @@
...
+    public ImmutableBiMap<K, V> toImmutable()
+    {
+        synchronized (this.lock)
+        {
+            return BiMaps.immutable.withAll(this);
+        }
+    }
+
+    /**
+     * @deprecated in 7.1. Use {@link OrderedIterable#zip(Iterable)} instead.
```
Comment (itohro) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/SynchronizedMutableBiMap.java]: Change version# to 8.0, and link to `zipWIthIndex()`.

**Comment by itohro** [comment]
Similar to previous change, you can put your name in copyright section in the files you created. Also modify other files to "Goldman Sachs and others.".

**Comment by bhav0904** [comment]
I worked on all your review comments, Hiroshi. But the ip validaton fails because it includes Craig's commit. We have to wait for @motlin to either cherry-pick my commit and push it along with his changes, or create a separate pull request for his changes.

**Comment by bhav0904** [comment]
I'm going to include another commit in the same pull request that deprecates all the zip/zipWithIndex methods in the BiMap hierarchy. Will commit tomorrow.

**Comment by itohro** [comment]
Thx, the latest changes look good to me.

If @motlin is also fine, we can merge it. We can manually validate your CLA so it is fine.

**Comment by motlin** [comment]
I included the cherry-picked commit in #96. That's a big pull request but if it's possible to get that merged in the near future then we could rebase this on top, have a simple history, and no CLA issue. I think it's fine to merge in either order though.

**Comment by itohro** [comment]
#96 has been merged. @bhav0904 , could you please rebase your 2 commits on top, then hopefully it resolves ip validation.

**Comment by bhav0904** [comment]
The build failed due to a timeout issue. Is there a way we can re-run the build without pushing any code?

**Comment by itohro** [comment]
Yes, there is a re-run button (a round arrow button) on the right top side of your build.

**Comment by bhav0904** [comment]
I don't have the option, @itohro. Could you re-run it for me please?

**Comment by itohro** [comment]
Ah ok, I didn't know it's only visible to committers. I've kicked off the build.

**Comment by bhav0904** [comment]
Thanks. Looks like it has gone through!

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/SynchronizedMutableBiMap.java [hunk 5]
```diff
@@ -0,0 +1,478 @@
...
+import org.eclipse.collections.impl.collection.mutable.SynchronizedMutableCollection;
+import org.eclipse.collections.impl.factory.BiMaps;
+import org.eclipse.collections.impl.list.fixed.ArrayAdapter;
+import org.eclipse.collections.impl.map.AbstractSynchronizedMapIterable;
+import org.eclipse.collections.impl.map.mutable.SynchronizedBiMapSerializationProxy;
+import org.eclipse.collections.impl.map.mutable.SynchronizedMapSerializationProxy;
+import org.eclipse.collections.impl.set.mutable.SynchronizedMutableSet;
+import org.eclipse.collections.impl.utility.LazyIterate;
+
+public class SynchronizedMutableBiMap<K, V> extends AbstractSynchronizedMapIterable<K, V> implements MutableBiMap<K, V>, Serializable
```
Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/SynchronizedMutableBiMap.java]: I think this can just be named SynchronizedBiMap. The fact that it is mutable is implied. The reason we don't follow that naming convention for SynchronizedMutableList, etc. is that there are already classes named SynchronizedList built into the JDK. This would follow the naming convention of SynchronizedBag for example.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/SynchronizedMutableBiMap.java [hunk 6]
```diff
@@ -0,0 +1,478 @@
...
+    }
+
+    protected SynchronizedMutableBiMap(MutableBiMap<K, V> delegate, Object lock)
+    {
+        super(delegate, lock);
+    }
+
+    /**
+     * This method will take a MutableBiMap and wrap it directly in a SynchronizedMutableMap.  It will
+     * take any other non-EC-map and first adapt it will a MapAdapter, and then return a
```
Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/SynchronizedMutableBiMap.java]: This Javadoc is copied from elsewhere and not accurate for BiMaps.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/SynchronizedMutableBiMap.java [hunk 7]
```diff
@@ -0,0 +1,478 @@
...
+            return SynchronizedMutableBiMap.of(this.getDelegate().clone());
+        }
+    }
+
+    public MutableBiMap<K, V> tap(Procedure<? super V> procedure)
+    {
+        synchronized (this.lock)
+        {
+            this.forEach(procedure);
+            return this;
```
Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/SynchronizedMutableBiMap.java]: I think every method should delegate if possible.

```
return this.getDelegate().tap(procedure)
```
This class should be largely auto-generated by code --> generate --> delegate methods --> getDelegate().

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/SynchronizedMutableBiMap.java [hunk 8]
```diff
@@ -0,0 +1,478 @@
...
+        {
+            return this.getDelegate().newEmpty();
+        }
+    }
+
+    public MutableBiMap<V, K> inverse()
+    {
+        synchronized (this.lock)
+        {
+            return this.getDelegate().inverse();
```
Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/SynchronizedMutableBiMap.java]: This can't be safe. You can mutate this.delegate by just calling inverse() and mutating the inverse, no?

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/SynchronizedMutableBiMap.java [hunk 9]
```diff
@@ -389,7 +393,7 @@ public ShortIterable collectShort(ShortFunction<? super V> shortFunction)
     }
 
     /**
-     * @deprecated in 7.1. Use {@link OrderedIterable#zip(Iterable)} instead.
+     * @deprecated in 8.0. Use {@link OrderedIterable#zip(Iterable)} instead.
```
Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/SynchronizedMutableBiMap.java]: Don't understand this part. Wasn't the deprecation in 7.1?

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/SynchronizedMutableBiMap.java (hunk 7 above)
Comment (bhav0904) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/SynchronizedMutableBiMap.java]: This method doesn't delegate in the way you coded it is because it has to return a MutableBiMap. If we return this.getDelegate().tap(procedure), it will return a non-synchronized copy of the map - isn't that unacceptable? I delegated all methods with the exception of the ones that return "this". For example, withKeyValue(). 

Still, we can change this to:

this.getDelegate().tap(procedure);
return this;

Do you agree?

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/SynchronizedMutableBiMap.java (hunk 8 above)
Comment (bhav0904) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/SynchronizedMutableBiMap.java]: Agreed. I am going to change it to:

`return new SynchronizedBiMap<V, K>(this.getDelegate().inverse());`

Is that ok?

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/SynchronizedMutableBiMap.java (hunk 9 above)
Comment (bhav0904) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/SynchronizedMutableBiMap.java]: I coded this back when 7.1 wasn;t released, and I committed it that way. Later I changed everything to be 8.0. SynchronizedBiMap didn't exist in 7.1, and neither did the deprecation of these methods in the BiMap heirarchy.

**Comment by bhav0904** [comment]
I have the changes ready, will push them once I get the OK from @motlin on each of my comments. Thanks!

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/SynchronizedMutableBiMap.java (hunk 7 above)
Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/SynchronizedMutableBiMap.java]: Right, that makes sense.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/SynchronizedMutableBiMap.java (hunk 8 above)
Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/SynchronizedMutableBiMap.java]: I think it has to share the lock too.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/SynchronizedMutableBiMap.java (hunk 9 above)
Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/SynchronizedMutableBiMap.java]: Ok I understand now. The problem is that you have two commits in this pull request which aren't independent. The first one sets the deprecation to 7.1, which is where @itohro commented here https://github.com/eclipse/eclipse-collections/pull/92/commits/f1d2938ed7946f72bb35e3a120407b5252fc25fc#diff-1ae117ff066180d4b3d50fbfc12db191R392. The second commit changes it to 8.0. For a clean history, 7.1 should never appear in any commit.

**Comment by bhav0904** [comment]
I've made the changes that @motlin suggested.

**Comment by itohro** [comment]
Confirmed all changes are reflected.
